gql_ws_link/lib/src/link.dart

```diff
@@ -0,0 +1,59 @@
+import "dart:async";
+
+import "package:gql_exec/gql_exec.dart";
+import "package:gql_link/gql_link.dart";
+import "package:gql_ws_link/src/get_from_ast.dart";
+import "package:gql_ws_link/src/socket_client.dart";
+import "package:gql_ws_link/src/messages.dart";
+import "package:gql/language.dart";
+import "package:gql/ast.dart";
+
+/// A Universal WebSocket [Link] implementation to support the WebSocket-GraphQL transport.
+/// It supports subscriptions, query and mutation operations as well.
+///
+/// NOTE: the actual socket connection will only get established after a [Request] is handled by this [WSLink].
+/// If you'd like to connect to the socket server instantly, call the [connectOrReconnect] method after creating this [WSLink] instance.
+class WSLink extends Link {
+  final String url;
+  SocketClient _client;
+  final SocketClientConfig config;
+
+  ResponseParser parser;
+
+  WSLink(this.url, {this.config, this.parser = const ResponseParser()});
```

Comment:
It's best to make `WSLink` to receive custom headers (and certificates if possible), as some servers require custom authentication, just like HTTPLink does: https://github.com/gql-dart/gql/blob/master/gql_http_link/lib/src/link.dart#L73-L74

Reply:
There’s connectionParams for that.
Problem with headers, is that browser-environments don’t allow customizing then when performing the WebSocket HTTP upgrade request

Reply:
hmm, I'm not familiar with websockets, but is there some kind of pattern on authentication then? (which works on the browser)

Reply:
You can send custom payload upon the socket creation...
In Apollo's case they used something called "connectionParams" which is far less standardised than HTTP headers but it's fine